File: packages/service/src/domain/songs/songs-repository.ts
import { Song } from '@service/domain/songs/song';
import { Repository } from '@service/infrastructure/repository';
import { DatabaseClient } from '@service/infrastructure/storage';
import { Service } from 'typedi';

@Service()
class SongsRepository extends Repository<Song> {
  private constructor(storage: DatabaseClient) {
    super('songs', storage);
  }

  findAllOrderedByTimesPlayedDesc(limit: number = 0): Promise<Song[]> {
    return this.collection.find().sort({ timesPlayed: -1 }).limit(limit).toArray();
  }

  findByName(name: string): Promise<Song | null> {
    return this.collection.findOne({ name });
  }

  findByYoutubeId(youtubeId: string): Promise<Song | null> {
    return this.collection.findOne({ youtubeId });
  }

  findByYoutubeIds(youtubeIds: string[]): Promise<Song[]> {
    return this.collection.find({ youtubeId: { $in: youtubeIds } }).toArray();
  }

  async insert(song: Song): Promise<void> {
    await this.collection.insertOne(song);
  }

  async insertMany(songs: Song[]): Promise<void> {
    await this.collection.insertMany(songs);
  }

  async replace(song: Song): Promise<void> {
    await this.collection.replaceOne({ _id: song._id }, song);
  }
}

export { SongsRepository };

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/songs/SongsRepository.kt
package xyz.lebkuchenfm.domain.songs

import com.github.michaelbull.result.Result
import xyz.lebkuchenfm.domain.youtube.YoutubeVideoId

interface SongsRepository {
    suspend fun insert(song: Song): Result<Song, InsertSongError>
    suspend fun findAllOrderByNameAsc(): List<Song>
    suspend fun findByName(name: String): Song?
    suspend fun findByYoutubeId(youtubeId: YoutubeVideoId): Song?
    suspend fun findByYoutubeIds(youtubeIds: List<YoutubeVideoId>): List<Song>
    suspend fun findRandom(limit: Int, phrase: String): List<Song>
    suspend fun incrementPlayCountByName(name: String): Song?
}

sealed class InsertSongError {
    data object UnknownError : InsertSongError()
}
